### PR TITLE
#1226 - Convert cartesian product of zonotopes to a new zonotope 

### LINKED ISSUE #1226: Convert cartesian product of zonotopes to a zonotope
```julia
julia> Z1 = rand(Zonotope, dim=1);
julia> Z2 = rand(Zonotope, dim=2);

julia> Z1 × Z2
CartesianProduct{Float64,Zonotope{Float64},Zonotope{Float64}}(Zonotope{Float64}([-0.00689418], [0.495163]), Zonotope{Float64}([-0.155357, 1.26408], [1.29322 0.20847 0.1
06569 0.87991; 1.39914 -0.466377 -0.093762 0.625436]))

julia> convert(Zonotope, Z1 × Z2)
ERROR: MethodError: Cannot `convert` an object of type CartesianProduct{Float64,Zonotope{Float64},Zonotope{Float64}} to an object of type Zonotope
Closest candidates are:
  convert(::Type{#s16} where #s16<:Union{Number, T}, ::MultivariatePolynomials.AbstractPolynomialLike{T}) where T at /Users/forets/.julia/dev/MultivariatePolynomials/sr
c/conversion.jl:10
  convert(::Type{Zonotope}, ::AbstractHyperrectangle) at /Users/forets/.julia/dev/LazySets/src/convert.jl:257
  convert(::Type{Zonotope}, ::CartesianProduct{N<:Real,#s66,#s65} where #s65<:AbstractHyperrectangle{N<:Real} where #s66<:AbstractHyperrectangle{N<:Real}) where N<:Real
 at /Users/forets/.julia/dev/LazySets/src/convert.jl:276
  ...
Stacktrace:
 [1] top-level scope at none:0
```

### PR DESCRIPTION
**Pull request by mforets**
Closes #1226.

### CODE REVIEW

File: src/convert.jl [hunk 1]
```diff
@@ -455,6 +455,36 @@ function convert(::Type{Hyperrectangle},
      return Hyperrectangle(low=l, high=h)
 end
 
+"""
+    convert(::Type{Zonotope}, cp::CartesianProduct{N, Zonotope{N}, Zonotope{N}}) where {N<:Real}
+
+Converts the cartesian product of two zonotoes to a new zonotope.
```
Comment (schillic) [inline src/convert.jl]:
```suggestion
Converts the cartesian product of two zonotopes to a new zonotope.
```

File: src/convert.jl [hunk 2]
```diff
@@ -455,6 +455,36 @@ function convert(::Type{Hyperrectangle},
...
+- `Zonotope` -- type used for dispatch
+- `S`        -- cartesian product of two zonotopes
+
+### Output
+
+A zonotopes.
+
+### Algorithm
+
+This implementation consists of concatenating the centers of each input zonotope.
```
Comment (schillic) [inline src/convert.jl]:
```suggestion
This implementation concatenates the centers of each input zonotope.
```

File: src/convert.jl [hunk 3]
```diff
@@ -455,6 +455,36 @@ function convert(::Type{Hyperrectangle},
...
+The resulting generator matrix is such that the generators for each element of the
+cartesian product are added along the diagonal.
+"""
+function convert(::Type{Zonotope}, cp::CartesianProduct{N, Zonotope{N}, Zonotope{N}}) where {N<:Real}
+    Z1, Z2 = cp.X, cp.Y
+    nZ1, nZ2 = size(Z1.generators, 1), size(Z2.generators, 1)
+    mZ1, mZ2 = size(Z1.generators, 2), size(Z2.generators, 2)
+    c = vcat(Z1.center, Z2.center)
+    G = [Z1.generators zeros(nZ1, mZ2);
+         zeros(nZ2, mZ1) Z2.generators]
```
Comment (schillic) [inline src/convert.jl]: Maybe consider a creating a sparse matrix. If the original generators are sparse matrices, we can also use [`SparseArrays.blockdiag`](https://docs.julialang.org/en/v1/stdlib/SparseArrays/#SparseArrays.blockdiag).

File: src/convert.jl [hunk 4]
```diff
@@ -455,6 +455,36 @@ function convert(::Type{Hyperrectangle},
...
+### Algorithm
+
+This implementation consists of concatenating the centers of each input zonotope.
+The resulting generator matrix is such that the generators for each element of the
+cartesian product are added along the diagonal.
+"""
+function convert(::Type{Zonotope}, cp::CartesianProduct{N, Zonotope{N}, Zonotope{N}}) where {N<:Real}
+    Z1, Z2 = cp.X, cp.Y
+    nZ1, nZ2 = size(Z1.generators, 1), size(Z2.generators, 1)
+    mZ1, mZ2 = size(Z1.generators, 2), size(Z2.generators, 2)
```
[the remark this answers is not in the document]
Comment (mforets) [reply src/convert.jl]: Yes :) i had in mind `n` for the dimension and `m` for the number of generators, but now i see it is confusing

File: src/convert.jl (hunk 3 above)
Comment (mforets) [reply src/convert.jl]: True. Unfortunately, `Zonotope` is not (yet?) parameterized in the center and generator's type. I propose to keep it as is, until we can use `blockdiag` consistently with the type of the zonotopes.

Comment (schillic) [reply src/convert.jl]: And what about the first point? The matrix will be sparse (>= 50% will be zeros).

Comment (mforets) [reply src/convert.jl]: Good, let's make `G` sparse.

Comment (mforets) [reply src/convert.jl]: `blockdiag(sparse(Z1.generators), sparse(Z2.generators))` should do it.

File: test/unit_Zonotope.jl [hunk 5]
```diff
@@ -110,7 +110,7 @@ for N in [Float64, Rational{Int}, Float32]
     Z1 = Zonotope(N[0], hcat(N[1]))
     Z2 = Zonotope(N[1/2], hcat(N[1/2]))
     Z = convert(Zonotope, Z1×Z2)
-    @test Z isa Zonotope && Z.center == N[0, 1/2] && Z.generators == N[1 0; 0 1/2]
+    @test Z isa Zonotope && Z.center == N[0, 1/2] && diag(Z.generators) == N[1, 1/2]
```
Comment (schillic) [inline test/unit_Zonotope.jl]: I think the old test would have been better. Now you don't test that the other entries are zeros.

Comment (mforets) [reply test/unit_Zonotope.jl]: ok, `Matrix(Z.generators) == N[1 0; 0 1/2]` should do it.

**Comment by schillic** [comment]
Sorry, `blockdiag` is not defined in v0.6 :frowning_face:

**Pull request merged**